src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -294,56 +372,77 @@ private TimeSpan GetTimeout()
         }
 
         // Internal for tests
-        internal void DeadlineExceeded(object state)
+        internal async void DeadlineExceeded(object? state)
         {
-            if (!_disposed)
+            // Deadline could be raised after call has been disposed
+            if (_deadlineDisposed)
             {
-                Debug.Assert(_deadlineTimer != null, "Should only be called via the timer.");
-                lock (_deadlineTimer)
-                {
-                    if (!_disposed)
-                    {
-                        Log.DeadlineExceeded(_logger, GetTimeout());
+                return;
+            }
 
-                        _status = new Status(StatusCode.DeadlineExceeded, "Deadline Exceeded");
+            Debug.Assert(_deadlineLock != null, "Lock has not been created.");
 
-                        try
-                        {
-                            // TODO(JamesNK): I believe this sends a RST_STREAM with INTERNAL_ERROR. Grpc.Core sends NO_ERROR
-                            HttpContext.Abort();
-                        }
-                        catch (Exception ex)
-                        {
-                            Log.DeadlineCancellationError(_logger, ex);
-                        }
-                    }
-                }
-            }
-        }
+            await _deadlineLock.WaitAsync();
 
-        public void Dispose()
-        {
-            var lockTaken = false;
             try
             {
-                // Only take a lock if there is a deadline
-                if (_deadlineTimer != null)
+                // Re-check after lock
+                if (_deadlineDisposed)
```

Comment:
But you just acquired the _deadlineLock, how could it be disposed?
Does _deadlineLock.WaitAsync throw if disposed?

Reply:
Do you have to release the SemaphoreSlim before disposing it?

I assume calls to SemaphoreSlim.WaitAsync() after disposal would throw an ODE, but do the ongoing calls awaiting SemaphoreSlim.WaitAsync() also see an ODE when it gets disposed?

I wonder if we need to add a Monitor now in order to avoid calling to SemaphoreSlim.WaitAsync() after we've disposed it.

Reply:
> But you just acquired the _deadlineLock, how could it be disposed?

Yeah, this line should be removed. It will never do anything. It should be impossible to be here after dispose because the timer is disposed of first with DisposeAsync.

Reply:
> I wonder if we need to add a Monitor now in order to avoid calling to SemaphoreSlim.WaitAsync() after we've disposed it.

I think I've already handled that situation. WaitAsync should never be called here because the timer is disposed of first with DisposeAsync.

Reply:
https://github.com/grpc/grpc-dotnet/blob/55d1d3f370b230c3bd38b05777d040b0eafa3c80/src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs#L425-L444